### ISSUE TITLE
if rollout.n is doubled, will the samples used for training doubled too?

### ISSUE THREAD

**Issue by StarDewXXX** [volcengine/verl#180]
if rollout.n is doubled, will the samples used for training doubled too? And does each forward/backward pass incur higher GPU costs in training because of larger rollout.n?

**Comment by vermouth1992** [comment]
Yes. Your understanding is correct.

**Comment by StarDewXXX** [comment]
> Yes. Your understanding is correct.

Hi, thanks for your reply. So if I increase rollout.n for m times, should I decrease ppo_mini_batch_size or ppo_micro_batch_size to 1/m of original value?

**Comment by vermouth1992** [comment]
> > Yes. Your understanding is correct.
> 
> Hi, thanks for your reply. So if I increase rollout.n for m times, should I decrease ppo_mini_batch_size or ppo_micro_batch_size to 1/m of original value?

I think ppo_mini_batch_size should also multiply by n. ppo_micro_batch_size should not because it only affects the gradient accumulation and increasing ppo_micro_batch_size may cause OOM

**Comment by JiaxingSong718** [comment]
> > > Yes. Your understanding is correct.
> > 
> > 
> > Hi, thanks for your reply. So if I increase rollout.n for m times, should I decrease ppo_mini_batch_size or ppo_micro_batch_size to 1/m of original value?
> 
> I think ppo_mini_batch_size should also multiply by n. ppo_micro_batch_size should not because it only affects the gradient accumulation and increasing ppo_micro_batch_size may cause OOM

ppo_micro_batch_size affect gradient accumulation, and what is the role of train_batch_size?

**Comment by splitnum11** [comment]
> > > > Yes. Your understanding is correct.
> > > 
> > > 
> > > Hi, thanks for your reply. So if I increase rollout.n for m times, should I decrease ppo_mini_batch_size or ppo_micro_batch_size to 1/m of original value?
> > 
> > 
> > I think ppo_mini_batch_size should also multiply by n. ppo_micro_batch_size should not because it only affects the gradient accumulation and increasing ppo_micro_batch_size may cause OOM
> 
> ppo_micro_batch_size affect gradient accumulation, and what is the role of train_batch_size?

same question

**Comment by asirgogogo** [comment]
same question

**Comment by Zoisaang** [comment]
I have the same issue and I want to describe it in more detail.
Here are the settings for GRPO in the example codes：

data.train_batch_size=1024
actor_rollout_ref.actor.ppo_mini_batch_size=256
actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=80
actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=160
actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=160
actor_rollout_ref.rollout.n=5

1. I’d like to know whether data.train_batch_size = 1024 refers to the number of prompts, or the number of prompts actor_rollout_ref.rollout.n. In other words，if batch data is 1024*5 or 1024?

2. actor_rollout_ref.actor.ppo_mini_batch_size=256, is it refers to the prompts or #prompts * actor_rollout_ref.rollout.n?

3. actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=80  cannot divide the actor_rollout_ref.actor.ppo_mini_batch_size=256  evenly — why is that?

4. Does the value of actor_rollout_ref.actor.ppo_mini_batch_size=256 and actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu = 80 consider actor_rollout_ref.rollout.n=5？That is to say, how do I calculate my training steps？Is 1024*5 / 256 or 1024 / 256？

**Comment by Zoisaang** [comment]
I have the same issue and I want to describe it in more detail.
Here are the settings for GRPO in the example codes：

data.train_batch_size=1024
actor_rollout_ref.actor.ppo_mini_batch_size=256
actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=80
actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=160
actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=160
actor_rollout_ref.rollout.n=5

1. I’d like to know whether data.train_batch_size = 1024 refers to #prompts, or #prompts * actor_rollout_ref.rollout.n. In other words，if batch data is 1024*5 or 1024?

2. actor_rollout_ref.actor.ppo_mini_batch_size=256, is it refers to #prompts or #prompts * actor_rollout_ref.rollout.n?

3. actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=80 cannot divide the actor_rollout_ref.actor.ppo_mini_batch_size=256 evenly — why is that? Is this a misconfiguration in the example code?

4. Does the value of actor_rollout_ref.actor.ppo_mini_batch_size=256 and actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu = 80 consider actor_rollout_ref.rollout.n=5？What I mean is that  how do I calculate my training steps？Is 1024*5 / 256 or 1024 / 256？And how do I calculate accumulated steps？Is 256 / 80 or 256*5/80？

🤗 I would appreciate it a lot if anyone could  help me clarify these questions. 🤗 🌻

**Comment by bestbzw** [comment]
same question

**Comment by Kirrito-k423** [comment]
English:

Based on the current code/docs, `data.train_batch_size` is the number of prompts sampled for one training iteration, not `#prompts * rollout.n`. After rollout, the number of response samples used by training becomes `data.train_batch_size * actor_rollout_ref.rollout.n`.

Evidence:
- The dataloader samples `batch_size=self.config.data.get("gen_batch_size", self.config.data.train_batch_size)`: https://github.com/verl-project/verl/blob/09de9858bd67e5359a043c59a381513477a989ec/verl/trainer/ppo/ray_trainer.py#L405-L408
- During training, the prompt batch is repeated by `rollout.n` before generation, and the original batch is repeated again to align with repeated responses: https://github.com/verl-project/verl/blob/09de9858bd67e5359a043c59a381513477a989ec/verl/trainer/ppo/ray_trainer.py#L1444-L1495
- The docs say each rollout produces `train_batch_size * n` samples: https://github.com/verl-project/verl/blob/09de9858bd67e5359a043c59a381513477a989ec/docs/perf/best_practices.rst#L142-L148

For `actor_rollout_ref.actor.ppo_mini_batch_size`: treat the configured value as prompt-level. The trainer multiplies it by `rollout.n` before actor update, so the effective mini-batch over response samples is `ppo_mini_batch_size * rollout.n`: https://github.com/verl-project/verl/blob/09de9858bd67e5359a043c59a381513477a989ec/verl/trainer/ppo/ray_trainer.py#L1307-L1318

So with:
- `data.train_batch_size=1024`
- `actor_rollout_ref.rollout.n=5`
- `actor_rollout_ref.actor.ppo_mini_batch_size=256`

the rollout produces `1024 * 5 = 5120` response samples, and the effective actor mini-batch size is `256 * 5 = 1280` response samples. That gives `5120 / 1280 = 4` actor mini-batches per PPO epoch, equivalently `1024 / 256 = 4` at the prompt level.

For `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu`: this is a local per-GPU forward/backward chunk size. It controls memory/gradient accumulation and is not a prompt-level sampling parameter. The docs describe it as samples processed per forward pass on one GPU group, and the worker code passes it into the engine as `micro_batch_size_per_gpu`: https://github.com/verl-project/verl/blob/09de9858bd67e5359a043c59a381513477a989ec/docs/perf/best_practices.rst#L145-L148 and https://github.com/verl-project/verl/blob/09de9858bd67e5359a043c59a381513477a989ec/verl/workers/engine_workers.py#L333-L340

Therefore `ppo_micro_batch_size_per_gpu=80` not dividing the configured `ppo_mini_batch_size=256` is not by itself a misconfiguration. In this example the effective global mini-batch is `256 * 5 = 1280`; the actual number of micro-batches depends on data-parallel size and whether dynamic batch sizing is enabled. For non-dynamic batching, the local data chunk must be divisible by `micro_batch_size_per_gpu`: https://github.com/verl-project/verl/blob/09de9858bd67e5359a043c59a381513477a989ec/verl/workers/engine/utils.py#L87-L95

中文：

根据当前代码和文档，`data.train_batch_size` 表示一次训练 iteration 采样的 prompt 数，不是 `prompt 数 * rollout.n`。rollout 之后，实际进入训练的 response sample 数会变成 `data.train_batch_size * actor_rollout_ref.rollout.n`。

对于 `actor_rollout_ref.actor.ppo_mini_batch_size`，可以按 prompt 维度理解配置值。trainer 在 actor update 前会把它乘以 `rollout.n`，所以 response sample 维度上的有效 mini-batch 是 `ppo_mini_batch_size * rollout.n`。

因此在 `train_batch_size=1024`、`rollout.n=5`、`ppo_mini_batch_size=256` 时，一轮 rollout 得到 `1024 * 5 = 5120` 条 response samples；actor 的有效 mini-batch 是 `256 * 5 = 1280` 条 response samples；每个 PPO epoch 的 actor mini-batch 数是 `5120 / 1280 = 4`，也就是 prompt 维度的 `1024 / 256 = 4`。

`ppo_micro_batch_size_per_gpu` 则是每个 GPU group 每次 forward/backward 的局部 micro-batch，用来控制显存和梯度累积，不是 prompt 采样数。它不需要直接整除配置里的 `ppo_mini_batch_size=256`；要看乘上 `rollout.n` 后的有效 mini-batch、data parallel size，以及是否启用 dynamic batch sizing。非 dynamic batch 下，本地 chunk 需要能被 `micro_batch_size_per_gpu` 整除。

So increasing `rollout.n` increases the number of generated/trained response samples. To keep the same prompt-level mini-batch behavior, do not divide `ppo_mini_batch_size` by `n`; the current trainer already scales the effective response-level mini-batch by `n`. Tune `ppo_micro_batch_size_per_gpu` mainly for memory/throughput.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
